File: producers/scala-producer/src/main/scala/com/mjarosie/ProducerDemoWithCallback.scala
package com.mjarosie

import org.apache.kafka.clients.producer._
import org.apache.kafka.common.serialization.StringSerializer
import org.slf4j.LoggerFactory

import java.util.Properties

object ProducerDemoWithCallback extends App {
  val logger = LoggerFactory.getLogger(this.getClass)

  val props = new Properties()
  props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka:9092")
  props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  val producer = new KafkaProducer[String, String](props)

  (1 to 10).map(i => {
    val record = new ProducerRecord[String, String]("first_topic", s"Hello world from Scala producer, message #${i}")
    producer.send(record, new Callback {
      override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit =
        if (exception != null) logger.error("Error while producing", exception)
        else logger.info(
          s"""Received callback after sending a message."
        Topic: ${metadata.topic}
        Partition: ${metadata.partition}
        Offset: ${metadata.offset}
        Timestamp: ${metadata.timestamp}""")
    })
  })

  producer.flush()
  producer.close()
}

File: producers/java-producer/src/main/java/com/mjarosie/ProducerDemoWithCallback.java
package com.mjarosie;

import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

/**
 * Hello world!
 *
 */
public class ProducerDemoWithCallback
{
    public static void main( String[] args )
    {
        final Logger logger = LoggerFactory.getLogger(ProducerDemoWithCallback.class);

        Properties properties = new Properties();
        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka:9092");
        properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());

        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(properties);

        for (int i =0; i < 10; i++) {
            final ProducerRecord<String, String> record = new ProducerRecord<>("first_topic", "Hello world from Java producer, message #" + i);
            producer.send(record, new Callback() {
                @Override
                public void onCompletion(RecordMetadata recordMetadata, Exception e) {
                    if (e != null) {
                        logger.error("Error while producing", e);
                    } else {
                        logger.info("Received callback after sending a message.\n" +
                                "Topic: " + recordMetadata.topic() + "\n" +
                                "Partition: " + recordMetadata.partition() + "\n" +
                                "Offset: " + recordMetadata.offset() + "\n" +
                                "Timestamp: " + recordMetadata.timestamp()
                        );
                    }
                }
            });
        }

        producer.flush();

        producer.close();
    }
}
